src/framework/execution/jobState.ts

```diff
@@ -1,16 +1,45 @@
 import { JobState, IntegrationStep } from './types';
 
 import { FileSystemGraphObjectStore } from '../storage';
+import { IntegrationDuplicateKeyError } from './error';
+import { Entity, Relationship } from '../types';
+
+export class DuplicateKeyTracker {
+  keySet = new Set<string>();
```

Comment:
Consider making this a `private readonly` property to prevent accidental direct access. We should always force usage of `registerKey`

Reply:
Good idea, will change